### ISSUE TITLE
Maybe offset-by-one error in compilation error message

### ISSUE THREAD

**Issue by qwfy** [rust-lang/rust#99049]
The following (maybe incorrect) error message is given by the compiler:

```
71  | trait ColumnValue: for<'r> sqlx::Encode<'r, Postgres> + sqlx::Type<Postgres> {
    |       ----------- this trait cannot be made into an object...
    = help: consider turning `type_info` into a method by giving it a `&self` argument or constraining it so it does not apply to trait objects
    = help: consider turning `compatible` into a method by giving it a `&self` argument or constraining it so it does not apply to trait objects
help: consider turning `type_info` into a method by giving it a `&self` argument
    |
175 |     fn type_info&self() -> DB::TypeInfo;
    |                 +++++ <======== (************ LOOK HERE ************)
help: consider turning `compatible` into a method by giving it a `&self` argument
    |
184 |     fn compatible&self, (ty: &DB::TypeInfo) -> bool {
    |                  ++++++
```

when compiling the code:

```
trait ColumnValue: for<'r> sqlx::Encode<'r, sqlx::Postgres> + sqlx::Type<sqlx::Postgres> {
}
```

Shouldn't the message labeled "LOOK HERE" be `fn type_info(&self) -> DB::TypeInfo;` ?

```
$ cargo --version
cargo 1.57.0 (b2e52d7ca 2021-10-21)

$ rustc --version
rustc 1.57.0 (f1edd0429 2021-11-29)
```

**Comment by compiler-errors** [comment]
Hi @qwfy, this is an old version of Rust. Do you mind checking if this reproduces on the current version of 1.62? Also, is the code sample you provided complete? It doesn't really look related to the errors you provided.

**Comment by qwfy** [comment]
@compiler-errors Yes, it is still present in 1.62. I've created a cargo project to reproduce the issue (just 3 lines of code).

The cargo project: https://github.com/qwfy/bug-reports/tree/master/gh99049-rust-offset-one-error-msg

The error message: https://github.com/qwfy/bug-reports/blob/f054a4effb94f95d5c724a9664a5fa92d98d01b2/gh99049-rust-offset-one-error-msg/src/lib.rs#L36

**Comment by compiler-errors** [comment]
Thanks for the reproduction, the sample you provided in "when compiling the code" in the issue above was not complete, but the cargo project you linked was. Anyways, I fixed this already in #98506, which is currently in nightly. The new output renders without those faulty suggestions.
